module/x/gravity/keeper/batch.go

```diff
@@ -1,300 +1,362 @@
+	info, found := k.GetEthereumInfo(ctx)
+	if !found {
+		return 0, sdkerrors.Wrap(
+			sdkerrors.ErrInvalidHeight,
+			"tracked ethereum height is 0. Track an populate the heights through a deposit event",
+		)
 	}
-	// we project how long it has been in milliseconds since the last Ethereum block height was observed
-	projectedMillis := (uint64(currentCosmosHeight) - heights.CosmosBlockHeight) * params.AverageBlockTime
-	// we convert that projection into the current Ethereum height using the average Ethereum block time in millis
-	projectedCurrentEthereumHeight := (projectedMillis / params.AverageEthereumBlockTime) + heights.EthereumBlockHeight
-	// we convert our target time for block timeouts (lets say 12 hours) into a number of blocks to
-	// place on top of our projection of the current Ethereum block height.
-	blocksToAdd := params.TargetBatchTimeout / params.AverageEthereumBlockTime
-	return projectedCurrentEthereumHeight + blocksToAdd
+
+	// calculate the time duration difference between the current block timestamp and the timestamp
+	// when the last Ethereum block height was observed on the bridge
+	timestampDiff := sdk.NewDec(int64(ctx.BlockTime().Sub(info.Timestamp)))
+
+	newBlocks := timestampDiff.QuoInt64(int64(params.AverageBlockTime)).TruncateInt64()
+	currentEthereumHeight := uint64(newBlocks) + info.Height
+
+	// TODO: [IMPORTANT] ensure timeout is in blocks and not time/duration on the contract
+	timeout := currentEthereumHeight + params.TargetBatchTimeout
+	return timeout, nil
 }
 
-// OutgoingTxBatchExecuted is run when the Cosmos chain detects that a batch has been executed on Ethereum
+// OnBatchTxExecuted is run when the Cosmos chain detects that a batch has been executed on Ethereum
 // It frees all the transactions in the batch, then cancels all earlier batches
-func (k Keeper) OutgoingTxBatchExecuted(ctx sdk.Context, tokenContract string, nonce uint64) error {
-	b := k.GetOutgoingTXBatch(ctx, tokenContract, nonce)
-	if b == nil {
-		return sdkerrors.Wrap(types.ErrUnknown, "nonce")
+func (k Keeper) OnBatchTxExecuted(ctx sdk.Context, tokenContract common.Address, batchTxID tmbytes.HexBytes) error {
+	batchTx, found := k.GetBatchTx(ctx, tokenContract, batchTxID)
+	if !found {
+		return sdkerrors.Wrapf(types.ErrTxNotFound, "batch tx for contract %s and tx id %s", tokenContract.String(), batchTxID.String())
 	}
 
-	// cleanup outgoing TX pool
-	for _, tx := range b.Transactions {
-		k.removePoolEntry(ctx, tx.Id)
+	// cleanup outgoing transfer tx pool
+	for _, transferTxID := range batchTx.Transactions {
+		k.DeleteTransferTx(ctx, transferTxID)
 	}
-	var err error
-	// Iterate through remaining batches
-	k.IterateOutgoingTXBatches(ctx, func(key []byte, iter_batch *types.OutgoingTxBatch) bool {
+
+	// Iterate through remaining batches with the same token contract
+	// TODO: still not getting why we need to cancel the other batch txs?
```

Comment:
It is impossible for an earlier batch to be executed because its nonce is too low

Reply:
This was also discussed here: https://blog.althea.net/how-gravity-works

>  After EthBlockDelay blocks on Ethereum, the validators submit the event to the Gravity module. When this event is fully observed (as detailed above), the batch is removed from the batch pool, since it is complete. Any batches of the same token with a lower batch nonce are also removed from the batch pool, but their transactions are also put back into the transaction pool. This is because we can be sure that it is impossible to submit them, since their batch nonces are too low. This mechanism allows transactions that were assembled into unprofitable batches to have a chance to make it into a profitable batch.